sqlx-mysql/src/connection/infile.rs

```diff
@@ -0,0 +1,385 @@
+                Poll::Ready(Err(e)) => return Poll::Ready(Err(e)),
+                Poll::Pending => return Poll::Pending,
+            }
+        }
+        // If the code reaches here the flush was succesful
+        assert!(self.send.is_none());
+        let mut send = SendPacket::new(buf, self.stream.sequence_id);
+        self.stream.sequence_id = self.stream.sequence_id.wrapping_add(1);
+        // Try to poll the send future right now
+        match Pin::new(&mut send).poll_send(cx, self.stream.socket_mut()) {
+            Poll::Ready(Ok(())) => {
+                self.send = None;
+                Poll::Ready(Ok(buf.len()))
+            }
+            Poll::Ready(Err(e)) => Poll::Ready(Err(e)),
+            Poll::Pending => {
+                // The send cannot happen right now but we did take the buffer to be sent
+                // On the next write, the scheduled send will be polled again
+                self.send = Some(send);
+                Poll::Ready(Ok(buf.len()))
+            }
+        }
+    }
+
+    fn poll_flush(
+        mut self: Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> Poll<Result<(), std::io::Error>> {
+        let send = self.send.take();
+        if let Some(mut send) = send {
+            match Pin::new(&mut send).poll_send(cx, self.stream.socket_mut()) {
+                Poll::Ready(Ok(())) => {
+                    self.send = None;
+                    Poll::Ready(Ok(()))
+                }
+                Poll::Ready(Err(e)) => Poll::Ready(Err(e)),
+                Poll::Pending => {
+                    self.send = Some(send);
+                    Poll::Pending
+                }
+            }
+        } else {
+            Poll::Ready(Ok(()))
+        }
+    }
+
+    fn poll_shutdown(
+        self: Pin<&mut Self>,
+        _cx: &mut std::task::Context<'_>,
+    ) -> Poll<Result<(), std::io::Error>> {
+        std::task::Poll::Ready(Ok(()))
+    }
+}
+
+struct SendPacket {
+    buf: Vec<u8>,
+}
+
+impl SendPacket {
+    fn new(data: &[u8], sequence_id: u8) -> Self {
+        let mut buf = Vec::with_capacity(data.len() + 4);
+        buf.extend_from_slice(&(data.len() as u32).to_le_bytes()[..3]);
+        buf.push(sequence_id);
+        buf.extend_from_slice(data);
+        assert!(buf.len() <= MAX_MYSQL_PACKET_SIZE + 4);
+        assert!(buf.len() >= 4);
+        assert!(buf.len() == data.len() + 4);
+        Self { buf }
+    }
+
+    fn poll_send(
+        mut self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+        socket: &mut impl Socket,
+    ) -> std::task::Poll<futures_io::Result<()>> {
+        let this = &mut *self;
+
+        while !this.buf.is_empty() {
+            match socket.try_write(&mut this.buf) {
+                Err(e) if e.kind() == std::io::ErrorKind::WouldBlock => {
+                    ready!(socket.poll_write_ready(cx))?;
+                }
+                Ok(written) => {
+                    this.buf.drain(..written);
```

Comment:
This is going to copy contents of the buffer on every poll

Reply:
From the docs it doesn't seem like drain should copy the contents of the buffer, can you explain?

Reply:
`drain` removes specified range of elements. But `Vec<_>` cannot have holes, so when you remove elements, it will move remaining elements in order to fill the hole.

In this case, `buf` could be 1 MB for example. May be `socket` only managed to write 10 KB (`written = 10_000`). When we call `buf.drain(..10_000)` it'll remove the first 10_000 bytes from the vector by moving the remaining 990_000 bytes to the beginning of the vector.

Reply:
This is another reason not to reinvent the wheel with buffering here. The connection already has a buffer that doesn't have this problem.